### PR TITLE
Optimize rule lookup

### PR DESCRIPTION
**Pull request by juan-punchman**
Addresses #453 

# Optimize rule lookup
This somewhat alleviates the speed issues with the `find_rule` function when `USE_OS_TZDB` is not on. In the while loop in `find_rule`, there were some very expensive function calls that make the lookup very slow. I have optimized some of the logic to minimize the amount of calls to slow functions.

### Database IO
Every call of the `find_next_rule` function read the entire database from disk into memory. That function would be called n times in the `find_rule` function. Obviously the database should only be read once from disk per call to the `find_rule` function. I have added an argument in `find_next_rule` to take a reference to the rules database.

### to_sys calculations
At the top of the while loop in the `find_rule` function, there was a call to convert the mdt argument to epoch time. I found that this call is actually pretty computationally expensive. The only value that can change the system time for `mdt` is the `x.save`. What I have done is calculate a base value for tx once, then get the value of tx from that base once per loop. This cuts down a great deal of computation there.

# Benchmarks
Some simple benchmarks show that these changes that I've made improve the runtime for calculating the local time of a zone time. The test I'm preforming use google's gtest library. The test code converts a zoned time to local time n times, then displays the result times. 

For reference, these benchmarks were run on:
```
ROG Zephryus S GX701GWR
OS: arch linux (linux 5.4.10)
CPU: i7-9750H (12) @ 4.500GHz
GPU: NVIDIA GeForce RTX 2070 Mobile
RAM: 32 GB @ 2666 Ghz
```

### Before Optimization
```
[----------] 1 test from date
[ RUN      ] date.bench1
lookup: once      : time used 1048982 us
lookup: every call: time used 619290 us
[       OK ] date.bench1 (1668 ms)
[----------] 1 test from date (1668 ms total)
```

### After Optimization
```
[ RUN      ] date.bench1
lookup: once      : time used 849095 us
lookup: every call: time used 434023 us
[       OK ] date.bench1 (1284 ms)
```

# The benchmark code:

```
TEST( date, bench1 )
{
    using namespace std::chrono;
    using namespace date;

    int n = 10000;

    auto sw = StopWatch();

    // lookup for every call in tz db,,,
    {
        auto tpUTC = std::chrono::system_clock::now();
        auto const* tz = date::locate_zone( "America/New_York" ); // locate the timezone once

        for( int i = 0; i != n; ++i )
        {
            auto const zt  = make_zoned( tz, tpUTC );
            auto const tp  = zt.get_local_time(); // expensive call
            auto const dp  = floor<days>( tp );
            auto const ymd = year_month_day( dp );
        }
    }
    fmt::print( "lookup: once      : time used {} us\n", sw.elapsedUs() );
    sw.restart();

    {
        auto tpUTC = std::chrono::system_clock::now();

        for( int i = 0; i != n; ++i )
        {
            auto const zt  = make_zoned( "America/New_York", tpUTC ); // locate the timezone every time
            auto const tp  = zt.get_local_time();
            auto const dp  = floor<days>( tp );
            auto const ymd = year_month_day( dp );
        }
    }
    fmt::print( "lookup: every call: time used {} us\n", sw.elapsedUs() );
}
```

# TL;DR
This simple optimization sped up the local time conversion by about 23% to 42% (depending on zone lookup).

### CODE REVIEW

**Comment by HowardHinnant** [comment]
Sorry for the delay in looking at this.

Your comment above about the entire db being read into memory on each iteration is incorrect.  `get_tzdb()` returns a reference to a function local static.  The initialization of this static is what reads the db into memory.  On subsequent calls, `get_tzdb()` simply returns a reference to that same static without re-initializing it.

I'm testing date/test/tz_test/validate.cpp with and without this patch and measuring virtually no difference in the execution times on average.  The variance between runs is much greater than the difference between the average of the runs.  I've tested with `AUTO_DOWNLOAD` set to both 0 and 1 to eliminate network effects.

While testing does appear to confirm your patch is correct (validate.cpp outputs the same answers), I'm hesitant to make this change as I'm not showing a clear benefit.

**Comment by jmc1283** [comment]
Yes, you are right about `get_tzdb()` returning a reference. Nevertheless, only calling it once eliminates some unnecessary function calls.

Maybe I wasn't clear in my initial comment, the function that I have really optimized is `get_local_time`. Since the rule lookup is an n time loop, the speed increase will vary by what time zone you are looking up. Because of that I'm not sure that the validate.cpp are the best tests for this PR, as my profiling of that file shows that only about 1/3 - 1/4 of the work that the validate.cpp is `get_local_time`. Despite this, on my system I saw a pretty significant 200ms difference in validate.cpp's execution time when comparing the master branch to my fork. 

This optimization shows its value when the loop in find_rule has many iterations by reducing unused work for zones with particularly slow lookups (like mine America/New_York).
